File: python/leetcode/tpl/SegmentTree.py
# http://toughcoder.net/blog/2022/08/23/rmq-made-easy/
class SegmentTree:
    def __init__(self, nums):
        self.n = len(nums)
        # 构造满二叉树
        self.tree = [0] * (4 * self.n)
        self.build(nums, 1, 0, self.n - 1)

    def build(self, nums, node, start, end):
        if start == end:
            self.tree[node] = nums[start]
        else:
            # 执行整数除法并向下取整
            mid = (start + end) // 2
            self.build(nums, 2 * node, start, mid)
            self.build(nums, 2 * node + 1, mid + 1, end)
            self.tree[node] = self.tree[2 * node] + self.tree[2 * node + 1]

    def update(self, index, value):
        def update_node(node, start, end):
            if start == end:
                print(node, value)
                self.tree[node] = value
            else:
                # 执行整数除法并向下取整
                mid = (start + end) // 2
                if start <= index <= mid:
                    update_node(2 * node, start, mid)
                else:
                    update_node(2 * node + 1, mid + 1, end)
                self.tree[node] = self.tree[2 * node] + self.tree[2 * node + 1]

        update_node(1, 0, self.n - 1)

    def query(self, query_start, query_end):
        def query_range(node, start, end, query_start, query_end):
            if query_start > end or query_end < start:
                return 0
            if query_start <= start and query_end >= end:
                return self.tree[node]
            mid = (start + end) // 2
            left_sum = query_range(2 * node, start, mid, query_start, query_end)
            right_sum = query_range(2 * node + 1, mid + 1, end, query_start, query_end)
            return left_sum + right_sum

        return query_range(1, 0, self.n - 1, query_start, query_end)


# 测试
nums = [1, 2, 3]
seg_tree = SegmentTree(nums)
print("线段树结构创建成功")
#
# print("线段树查询结果：")
# print("nums[0:2] 的区间和为：", seg_tree.query(0, 2))
# print("nums[1:4] 的区间和为：", seg_tree.query(1, 4))

print("更新后，nums[1:4] 的区间和为：", seg_tree.query(0, 3))

seg_tree.update(0, 6)
print("更新后，nums[1:4] 的区间和为：", seg_tree.query(0, 3))

File: cpp/tpl/SegmentTree.cpp
//
// Created by sanyinchen on 7/26/25.
//
// 线段树
class SegmentTree {
private:
    vector<int> tree;
    int n;

    // 构建线段树
    void build(vector<int>& nums, int node, int l, int r) {
        if (l == r) {
            tree[node] = nums[l];
            return;
        }
        int mid = (l + r) / 2;
        build(nums, 2 * node, l, mid);
        build(nums, 2 * node + 1, mid + 1, r);
        tree[node] = tree[2 * node] + tree[2 * node + 1];
    }

    // 查询区间 [ql, qr] 的和
    int query(int node, int l, int r, int ql, int qr) {
        if (qr < l || ql > r) return 0; // 不相交
        if (ql <= l && r <= qr) return tree[node]; // 完全包含
        int mid = (l + r) / 2;
        return query(2 * node, l, mid, ql, qr) + query(2 * node + 1, mid + 1, r, ql, qr);
    }

    // 单点更新：将 index 的值改为 val
    void update(int node, int l, int r, int index, int val) {
        if (l == r) {
            tree[node] = val;
            return;
        }
        int mid = (l + r) / 2;
        if (index <= mid)
            update(2 * node, l, mid, index, val);
        else
            update(2 * node + 1, mid + 1, r, index, val);
        tree[node] = tree[2 * node] + tree[2 * node + 1];
    }

public:
    SegmentTree(vector<int>& nums) {
        n = nums.size();
        tree.resize(n * 4);
        build(nums, 1, 0, n - 1);
    }

    int query(int l, int r) {
        return query(1, 0, n - 1, l, r);
    }

    void update(int index, int val) {
        update(1, 0, n - 1, index, val);
    }
};
